packages/core/contracts/oracle/implementation/Staker.sol

```diff
@@ -196,31 +216,58 @@ abstract contract Staker is StakerInterface, Ownable, Lockable {
      * @notice Send accumulated rewards to the voter. Note that these rewards do not include slashing balance changes.
      * @return uint256 the amount of tokens sent to the voter.
      */
-    function withdrawRewards() public override nonReentrant() returns (uint256) {
-        _updateTrackers(msg.sender);
-        VoterStake storage voterStake = voterStakes[msg.sender];
+    function withdrawRewards() public returns (uint256) {
+        return _withdrawRewards(msg.sender);
+    }
+
+    function _withdrawRewards(address from) internal returns (uint256) {
+        _updateTrackers(from);
+        VoterStake storage voterStake = voterStakes[from];
 
         uint256 tokensToMint = voterStake.outstandingRewards;
         if (tokensToMint > 0) {
             voterStake.outstandingRewards = 0;
-            require(votingToken.mint(msg.sender, tokensToMint), "Voting token issuance failed");
-            emit WithdrawnRewards(msg.sender, tokensToMint);
+            address recipient = from != msg.sender ? address(this) : from;
+            require(votingToken.mint(recipient, tokensToMint), "Voting token issuance failed");
+            emit WithdrawnRewards(from, msg.sender, tokensToMint);
         }
         return tokensToMint;
     }
 
     /**
-     * @notice Stake accumulated rewards. This is just a convenience method that combines withdraw with stake in the
-     * same transaction.
-     * @dev this method requires that the user has approved this contract.
-     * @return uint256 the amount of tokens that the user is staking.
+     * @notice Stake accumulated rewards. This is merely a convenience mechanism that combines the voter's withdrawal and stake
+     *  in the same transaction if requested by a delegate or the voter himself.
+     * @dev this method requires that the msg.sender(voter or delegate) has approved this contract.
+     * @dev The rewarded tokens simply pass through the delegate's wallet before being staked on the voter's behalf.
+     *  The balance of the delegate remains unchanged.
+     * @return uint256 the amount of tokens that the voter is staking.
      */
-    function withdrawAndRestake() external returns (uint256) {
-        uint256 rewards = withdrawRewards();
-        stake(rewards);
+    function withdrawAndRestake() external override returns (uint256) {
+        address voter = getVoterFromDelegate(msg.sender);
+        uint256 rewards = _withdrawRewards(voter);
+        _stakeTo(address(this), voter, rewards);
```

Comment:
Does this work when called by the voter directly?

It looks like _withdrawRewards() sends funds to the voter when voter == msg.sender, but _stakeTo, when called with from == address(this), assumes it will use tokens that are already in the contract. 

Higher-level comment: the special-casing in _withdrawRewards and _stakeTo is a bit convoluted and confusing. Would it makes sense to break these into simpler methods?

Some options to simplify:
1. _withdrawRewards(voter, recipient) just sends the tokens to whoever you specify the recipient to be and doesn't care about the caller. _stakeTo(voter, amount, fundsAlreadyInContract), where fundsAlreadyInContract is a bool that tells you whether to pull funds from the voter or not. This is still a little convoluted, but I think it makes the interfaces more clear.
2. Just reimplement most of the _withdrawRewards and _stakeTo logic in withdrawAndRestake. More bytecode, but it would be cheaper on gas and potentially easier to follow.

Thoughts?

Reply:
ye, I think I like this. the prop in the `_withdrawRewards` called `from` is a bit confusing and what you describe will help with this as `_withdrawRewards(voter, recipient)` will give us the _voter_ we are pulling rewards _from_ and the _recipeint_ they are being sent to, removing the need for this line:
```
            address recipient = from != msg.sender ? address(this) : from;
```

Reply:
Thank you guys.
I love the first suggestion: `_withdrawRewards(voter, recipient)`. Makes a lot of sense and simplifies things.

I believe it already worked when calling `withdrawRewards` by voter as this [test](https://github.com/UMAprotocol/protocol/blob/fbf6cec28a70fc7cd0a6cc70205400e05cd49589/packages/core/test/oracle/Staker.js#L99) illustrates. But the suggestion is better.

I have updated with `_withdrawRewards(voter, recipient)`. 

Happy to make any of the other changes @mrice32  proposes if you feel it necessary. In my opinion I don't see too much complexity and there is no repetition of code, so I'm fine with it. But I've been looking at it for a while so I can be byassed.

Reply:
SGTM!